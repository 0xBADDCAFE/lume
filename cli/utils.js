import { existsSync } from "../deps/fs.js";
import lume from "../mod.js";
import { join, resolve } from "../deps/path.js";
import { brightGreen } from "../deps/colors.js";

/**
 * @return {Promise<*>} a lume instance - ready to build, run, etc.
 */
export async function buildSite(options, site) {
  options.root = resolve(Deno.cwd(), options.root);
  const configFile = join(options.root, options.config);

  if (!site) {
    if (existsSync(configFile)) {
      const mod = await import(`file://${configFile}`);
      site = mod.default;
    } else {
      site = lume();
    }
  }

  site.options.cwd = options.root;

  if (options.dev) {
    site.options.dev = options.dev;
  }

  if (options.location) {
    site.options.location = new URL(options.location);
  }

  if (options.src) {
    site.options.src = options.src;
  }

  if (options.dest) {
    site.options.dest = options.dest;
  }

  if (options["--"]) {
    site.options.flags = options["--"];
  }

  return site;
}

/**
 * @param command name of the command you're validating
 * @param args array of cli arguments (i.e. options._)
 * @param max the max number of args
 * @param min the min number of args
 */
export function validateArgsCount(command, args, max, min = 0) {
  if (args.length > max) {
    throw new Error(`Unexpected arguments: ${args.slice(max).join(", ")}
    
    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }

  if (args.length < min) {
    throw new Error(`Missing arguments

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }
}
