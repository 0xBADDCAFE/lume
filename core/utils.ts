import { DOMParser, HTMLDocument } from "../deps/dom.ts";
import {
  brightGreen,
  brightYellow,
  cyan,
  dim,
  green,
  red,
} from "../deps/colors.ts";
import { dirname, extname, join, posix, SEP } from "../deps/path.ts";
import { parse } from "../deps/jsonc.ts";
import { Exception } from "./errors.ts";

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "imagick",
  "inline",
  "jsx",
  "jsx_preact",
  "katex",
  "lightningcss",
  "liquid",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "netlify_cms",
  "on_demand",
  "pagefind",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "remark",
  "resolve_urls",
  "sass",
  "sheets",
  "slugify_urls",
  "svgo",
  "terser",
  "windi_css",
];

/** A list of the available plugins with init configurations */
export const initPlugins = [
  "jsx",
  "jsx_preact",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

export function promptConfigUpdate(data: unknown) {
  log(
    red("deno.jsonc needs to be manually updated:"),
    dim("Use deno.json to update it automatically"),
    JSON.stringify(data, null, 2),
  );
}

/** Check the compatibility with the current Deno version */
export function checkDenoVersion(): void {
  const minimum = "1.25.4";
  const current = Deno.version.deno;

  if (current < minimum) {
    console.log("----------------------------------------");
    console.error(red("Your Deno version is not compatible with Lume"));
    console.log(`Lume needs Deno ${green(minimum)} or greater`);
    console.log(`Your current version is ${red(current)}`);
    console.log(`Run ${cyan("deno upgrade")} and try again`);
    console.log("----------------------------------------");
    Deno.exit(1);
  }
}

/** Check if the current version is outdated */
export async function checkUpgrade(): Promise<void> {
  const current = getLumeVersion();

  // It's a local version
  if (current.startsWith("local ")) {
    return;
  }

  const stable = !!current.match(/^v\d+\./);
  const expires = 1000 * 60 * 60 * 24; // 1 day
  const interval = localStorage.getItem("lume-upgrade");

  if (interval && parseInt(interval) + expires > Date.now()) {
    return;
  }

  localStorage.setItem("lume-upgrade", Date.now().toString());

  const latest = stable
    ? await getLatestVersion()
    : await getLatestDevelopmentVersion();

  if (current === latest) {
    return;
  }

  let global = false;
  try {
    await Promise.any([
      Deno.stat("deno.json"),
      Deno.stat("deno.jsonc"),
    ]);
    await Deno.stat("import_map.json");
  } catch {
    global = true;
  }

  const command = global
    ? (stable ? "lume upgrade --global" : "lume upgrade --dev --global")
    : (stable ? "deno task lume upgrade" : "deno task lume upgrade --dev");

  log(
    `Update available ${dim(current)}  â†’ ${green(latest)}`,
    `Run ${cyan(command)} to update`,
  );
}

/** Return the latest stable version from the deno.land/x repository */
export async function getLatestVersion(): Promise<string> {
  const response = await fetch("https://cdn.deno.land/lume/meta/versions.json");
  const versions = await response.json();
  return versions.latest;
}

/** Return the hash of the latest commit from the GitHub repository */
export async function getLatestDevelopmentVersion(): Promise<string> {
  const response = await fetch(
    "https://api.github.com/repos/lumeland/lume/commits?per_page=1",
  );
  const commits = await response.json();
  return commits[0].sha;
}

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx";
    jsxImportSource?: string;
  };
  [key: string]: unknown;
}

/** Run a callback concurrently with all the elements of an Iterable */
export async function concurrent<Type>(
  iterable: AsyncIterable<Type> | Iterable<Type>,
  iteratorFn: (arg: Type) => Promise<unknown>,
  limit = 200,
) {
  const executing: Promise<unknown>[] = [];

  for await (const item of iterable) {
    const p: Promise<unknown> = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

const decoder = new TextDecoder();
const encoder = new TextEncoder();

/** Encode a message using SHA-1 algorithm */
export async function sha1(message: string | Uint8Array): Promise<string> {
  if (typeof message === "string") {
    message = encoder.encode(message);
  }

  const hash = await crypto.subtle.digest("SHA-1", message);
  return decoder.decode(hash);
}

/** Helper to create optional properties recursively */
// deno-lint-ignore ban-types
export type DeepPartial<T> = T extends object ? {
    [P in keyof T]?: DeepPartial<T[P]>;
  }
  : T;

/**
 * Merge two objects recursively.
 * It's used to merge user options with default options.
 */
export function merge<Type>(
  defaults: Type,
  user?: Partial<Type> | DeepPartial<Type>,
): Type {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (value === undefined) {
      continue;
    }

    // @ts-ignore: No index signature with a parameter of type 'string' was found on type 'unknown'
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      // @ts-ignore: Type 'string' cannot be used to index type 'Type'
      merged[key] = merge(merged[key], value);
      continue;
    }

    // @ts-ignore: Type 'string' cannot be used to index type 'Type'
    merged[key] = value;
  }

  return merged;
}

const reactElement = Symbol.for("react.element");
const objectConstructor = {}.constructor;

/** Check if the argument passed is a plain object */
export function isPlainObject(obj: unknown): obj is Record<string, unknown> {
  return typeof obj === "object" && obj !== null &&
    obj.constructor === objectConstructor &&
    // @ts-ignore: Check if the argument passed is a React element
    obj["$$typeof"] !== reactElement;
}

/**
 * Convert the Windows paths (that use the separator "\")
 * to Posix paths (with the separator "/")
 * and ensure it starts with "/".
 */
export function normalizePath(path: string) {
  if (SEP !== "/") {
    path = path.replaceAll(SEP, "/");

    // Is absolute Windows path (C:/...)
    if (path.includes(":/")) {
      return path;
    }
  }

  return posix.join("/", path);
}

/** Convert an HTMLDocument instance to a string */
export function documentToString(document: HTMLDocument) {
  const { doctype, documentElement } = document;

  if (!doctype) {
    return `<!DOCTYPE html>\n${documentElement?.outerHTML || ""}`;
  }

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${documentElement?.outerHTML}`;
}

const parser = new DOMParser();

/** Parse a string with HTML code and return an HTMLDocument */
export function stringToDocument(string: string): HTMLDocument {
  const document = parser.parseFromString(string, "text/html");

  if (!document) {
    throw new Error("Unable to parse the HTML code");
  }

  return document;
}

/** Return the current installed version */
export function getLumeVersion(
  url = new URL(import.meta.resolve("../")),
): string {
  const { pathname } = url;
  return pathname.match(/@([^/]+)/)?.[1] ?? `local (${pathname})`;
}

/** Returns the _config file of a site */
export async function getConfigFile(
  path?: string,
): Promise<string | undefined> {
  if (path) {
    try {
      return await Deno.realPath(path);
    } catch (cause) {
      throw new Exception("Config file not found", { path, cause });
    }
  }

  const paths = ["_config.js", "_config.ts"];

  for (const path of paths) {
    try {
      return await Deno.realPath(path);
    } catch {
      // Ignore
    }
  }
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(
  importMapFile?: string,
): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      importMapFile ||= config.importMap;
      if (importMapFile) {
        config.importMap = importMapFile;
        const importMap: ImportMap = isUrl(importMapFile)
          ? await (await fetch(importMapFile)).json()
          : await JSON.parse(await Deno.readTextFile(importMapFile));

        return { file, config, importMap };
      }
      return { file, config };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}

/** Update the Deno configuration */
export async function writeDenoConfig(options: DenoConfigResult) {
  const { file, config, importMap } = options;

  if (importMap && !config.importMap) {
    config.importMap = "./import_map.json";
  }

  if (config.importMap) {
    const importMapFile = join(dirname(file), config.importMap);
    console.log(importMapFile);

    await Deno.writeTextFile(
      importMapFile,
      JSON.stringify(importMap, null, 2) + "\n",
    );
    console.log(brightGreen("Import map file saved:"), importMapFile);
  }

  if (extname(file) === ".jsonc") {
    console.log(
      brightYellow("deno.jsonc needs to be manually updated:"),
      "Use deno.json to update it automatically",
      JSON.stringify(config, null, 2),
    );
  } else {
    await Deno.writeTextFile(file, JSON.stringify(config, null, 2) + "\n");
    console.log(brightGreen("Deno configuration file saved:"), file);
  }
}

export function isUrl(path: string): boolean {
  return !!path.match(/^(https?|file):\/\//);
}

export function createDate(str: string | number): Date | undefined {
  if (typeof str === "number") {
    return new Date(str);
  }

  const datetime = str.match(
    /^(\d{4})-(\d\d)-(\d\d)(?:-(\d\d)-(\d\d)(?:-(\d\d))?)?$/,
  );

  if (datetime) {
    const [, year, month, day, hour, minute, second] = datetime;

    return new Date(Date.UTC(
      parseInt(year),
      parseInt(month) - 1,
      parseInt(day),
      hour ? parseInt(hour) : 0,
      minute ? parseInt(minute) : 0,
      second ? parseInt(second) : 0,
    ));
  }

  if (str.match(/^\d+$/)) {
    return new Date(parseInt(str));
  }
}

export async function read(path: string, isBinary: true): Promise<Uint8Array>;
export async function read(path: string, isBinary: false): Promise<string>;
export async function read(
  path: string,
  isBinary: boolean,
): Promise<string | Uint8Array> {
  if (!isUrl(path)) {
    return isBinary ? Deno.readFile(path) : Deno.readTextFile(path);
  }

  const response = await fetch(path);

  return isBinary
    ? new Uint8Array(await response.arrayBuffer())
    : response.text();
}
