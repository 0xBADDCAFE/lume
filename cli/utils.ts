import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { dim, red } from "../deps/colors.ts";
import { getConfigFile } from "../core/utils.ts";

import type { Site } from "../core.ts";

/** Create a site instance */
export async function createSite(root: string, config?: string): Promise<Site> {
  const path = await getConfigFile(root, config);

  if (path) {
    console.log(`Loading config file ${dim(path)}`);
    console.log();
    const mod = await import(toFileUrl(path).href);
    return mod.default;
  }

  return lume();
}

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "imagick",
  "inline",
  "jsx",
  "lightningcss",
  "liquid",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "nano_jsx",
  "netlify_cms",
  "on_demand",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "resolve_urls",
  "sass",
  "slugify_urls",
  "svgo",
  "terser",
  "windi_css",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

export function promptConfigUpdate(data: unknown) {
  log(
    red("deno.jsonc needs to be manually updated:"),
    dim("Use deno.json to update it automatically"),
    JSON.stringify(data, null, 2),
  );
}
