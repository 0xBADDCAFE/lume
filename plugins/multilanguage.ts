import { Page } from "../core/filesystem.ts";
import { isPlainObject, merge } from "../core/utils.ts";

import type { Data, Plugin } from "../core.ts";

export interface Options {
  /** The list of extensions used for this plugin */
  extensions: string[];

  /** The helper name */
  name: string;
}

// Default options
export const defaults: Options = {
  extensions: [".html"],
  name: "mergeLanguages",
};

type Alternates = Record<string, Page>;

export default function multilanguage(userOptions?: Partial<Options>): Plugin {
  const options = merge(defaults, userOptions);

  return (site) => {
    // Register the helper
    site.data(options.name, mergeLanguages);

    // Preprocessor to create new pages dynamically for each language
    site.preprocess(options.extensions, (page, pages) => {
      // The "lang" variable of the pages must be an array
      const languages = page.data.lang as string | string[] | undefined;

      if (!Array.isArray(languages)) {
        return;
      }

      // Create a Data for each language
      const languageData: Record<string, { data: Data; customUrl: boolean }> =
        {};
      languages.forEach((key) => {
        // deno-lint-ignore no-explicit-any
        const data: Record<string, any> = { ...page.data };

        // This language has a custom url (like url.en = "/english-url/")
        const customUrl = data[`url.${key}`] || data[key]?.url;

        // Remove all entries of other languages
        for (const [name, value] of Object.entries(data)) {
          if (languages.includes(name)) {
            if (name === key) {
              Object.assign(data, value);
            } else {
              delete data[name];
            }
          }
        }
        languageData[key] = {
          data: filterLanguage(languages, key, data),
          customUrl,
        };
      });

      const alternates: Alternates = {};

      // Create a new page per language
      const newPages: Page[] = [];

      for (const [l, { data, customUrl }] of Object.entries(languageData)) {
        data.alternates = alternates;
        data.lang = l;

        const newPage = page.duplicate(l);
        newPage.data = data;

        if (customUrl) {
          newPage.data.url = site.pagePreparer.getUrl(
            newPage,
            page.data.url as string,
          );
        } else {
          newPage.data.url = `/${l}${newPage.data.url}`;
        }

        alternates[l] = newPage;
        newPages.push(newPage);
      }

      // Replace the current page with the multiple languages
      pages.splice(pages.indexOf(page), 1, ...newPages);
    });

    // Include automatically the <link rel="alternate"> elements
    // with the other languages
    site.process(options.extensions, (page) => {
      const { document } = page;
      const alternates = page.data.alternates as
        | Alternates
        | undefined;
      const lang = page.data.lang as string | undefined;

      if (!document || !alternates || !lang) {
        return;
      }

      // Include <html lang="${lang}"> attribute element if it's missing
      if (!document.documentElement?.getAttribute("lang")) {
        document.documentElement?.setAttribute("lang", lang);
      }

      // Insert the <link> elements automatically
      for (const [altLang, altPage] of Object.entries(alternates)) {
        if (altLang === lang) {
          continue;
        }
        const meta = document.createElement("link");
        meta.setAttribute("rel", "alternate");
        meta.setAttribute("hreflang", altLang);
        meta.setAttribute("href", altPage.data.url);
        document.head.appendChild(meta);
        document.head.appendChild(document.createTextNode("\n"));
      }
    });
  };
}

/**
 * Manage multiple paginations from different languages.
 * Example:
 * ```ts
 * const pagination = paginateLanguages({
 *   en: paginate(englishPages),
 *   gl: paginate(galicianPages),
 * })
 * ```
 */
function mergeLanguages(
  pages: Record<string, Record<string, unknown>[]>,
): unknown[] {
  const result: unknown[] = [];
  const limit = Math.max(...Object.values(pages).map((v) => v.length));
  let index = 0;

  while (index < limit) {
    const page: Record<string, unknown> = {};

    for (const [lang, pageResults] of Object.entries(pages)) {
      const pageResult = pageResults[index];

      if (pageResult) {
        for (const [key, value] of Object.entries(pageResult)) {
          page[`${key}.${lang}`] = value;
        }
      }
    }
    result.push(page);
    index++;
  }

  return result;
}

/**
 * Remove the entries from all "langs" except the "lang" value
 */
function filterLanguage(langs: string[], lang: string, data: Data): Data {
  for (let [name, value] of Object.entries(data)) {
    if (isPlainObject(value)) {
      data[name] = value = filterLanguage(langs, lang, {
        ...value as Record<string, unknown>,
      });
    } else if (Array.isArray(value)) {
      data[name] = value = value.map((item) => {
        return isPlainObject(item)
          ? filterLanguage(langs, lang, { ...item as Record<string, unknown> })
          : item;
      });
    }

    const parts = name.match(/^(.*)\.([^.]+)$/);

    if (parts) {
      const [, key, l] = parts;

      if (langs.includes(l)) {
        if (lang === l) {
          data[key] = value;
        }

        delete data[name];
      }
    }
  }

  return data;
}
