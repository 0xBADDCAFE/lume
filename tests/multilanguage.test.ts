import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import multilanguage from "../plugins/multilanguage.ts";

Deno.test("multilanguage plugin", async (t) => {
  const site = getSite({
    src: "multilanguage",
  });

  site.use(multilanguage());

  await build(site);
  await assertSiteSnapshot(t, site);
});
