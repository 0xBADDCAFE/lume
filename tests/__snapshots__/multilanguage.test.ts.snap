export const snapshot = {};

snapshot[`multilanguage plugin 1`] = `9`;

snapshot[`multilanguage plugin 2`] = `
{
  formats: [
    {
      engines: 1,
      ext: ".tmpl.ts",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".tmpl.js",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: undefined,
      ext: ".tmpl.json",
      pageLoader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".json",
    },
    {
      engines: 1,
      ext: ".md",
      pageLoader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      pageLoader: [AsyncFunction: text],
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
    },
    {
      componentLoader: [AsyncFunction: text],
      engines: 1,
      ext: ".njk",
      includesPath: "_includes",
      pageLoader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yaml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yml",
      pageLoader: [AsyncFunction: yaml],
    },
  ],
}
`;

snapshot[`multilanguage plugin 3`] = `[]`;

snapshot[`multilanguage plugin 4`] = `
{
  content: "<!DOCTYPE html>
" +
    '<html lang="en"><head><link rel="alternate" hreflang="en" href="http://localhost/en/">
' +
    '<link rel="alternate" hreflang="gl" href="http://localhost/gl/">
' +
    "</head><body><p>Hello world</p>
" +
    "</body></html>",
  data: {
    alternates: [
      "en",
      "gl",
    ],
    children: true,
    content: "Hello world
",
    date: 1970-01-01T00:00:00.000Z,
    id: "index",
    lang: "en",
    mergedKeys: {
      en: "object",
      fr: "object",
      gl: "object",
      it: "object",
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [],
    title: "Hello world",
    url: "/en/",
  },
  dest: {
    ext: ".html",
    path: "/en/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/index",
    remote: undefined,
    slug: "index",
  },
}
`;

snapshot[`multilanguage plugin 5`] = `
{
  content: "<!DOCTYPE html>
" +
    '<html lang="gl"><head><link rel="alternate" hreflang="en" href="http://localhost/en/">
' +
    '<link rel="alternate" hreflang="gl" href="http://localhost/gl/">
' +
    "</head><body><p>Ola mundo</p>
" +
    "</body></html>",
  data: {
    alternates: [
      "en",
      "gl",
    ],
    children: true,
    content: "Ola mundo",
    date: 1970-01-01T00:00:00.000Z,
    id: "index",
    lang: "gl",
    mergedKeys: {
      en: "object",
      fr: "object",
      gl: "object",
      it: "object",
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [],
    title: "Ola mundo",
    url: "/gl/",
  },
  dest: {
    ext: ".html",
    path: "/gl/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/index",
    remote: undefined,
    slug: "index",
  },
}
`;

snapshot[`multilanguage plugin 6`] = `
{
  content: "<!DOCTYPE html>
" +
    '<html lang="en"><head><link rel="alternate" hreflang="en" href="http://localhost/en/other-page-english/">
' +
    '<link rel="alternate" hreflang="fr" href="http://localhost/fr/other-page-french.html">
' +
    "</head><body><p>Other page</p></body></html>",
  data: {
    alternates: [
      "en",
      "fr",
    ],
    children: true,
    content: "<p>Other page</p>",
    date: 1970-01-01T00:00:00.000Z,
    id: "other",
    lang: "en",
    mergedKeys: {
      en: "object",
      fr: "object",
      gl: "object",
      it: "object",
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [],
    title: "Other",
    url: "/en/other-page-english/",
  },
  dest: {
    ext: ".html",
    path: "/en/other-page-english/index",
  },
  src: {
    asset: undefined,
    ext: ".njk",
    path: "/other",
    remote: undefined,
    slug: "other",
  },
}
`;

snapshot[`multilanguage plugin 7`] = `
{
  content: "<!DOCTYPE html>
" +
    '<html lang="fr"><head><link rel="alternate" hreflang="en" href="http://localhost/en/other-page-english/">
' +
    '<link rel="alternate" hreflang="fr" href="http://localhost/fr/other-page-french.html">
' +
    "</head><body><p>Other page</p></body></html>",
  data: {
    alternates: [
      "en",
      "fr",
    ],
    children: true,
    content: "<p>Other page</p>",
    date: 1970-01-01T00:00:00.000Z,
    id: "other",
    lang: "fr",
    mergedKeys: {
      en: "object",
      fr: "object",
      gl: "object",
      it: "object",
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [],
    title: "Autre",
    url: "/fr/other-page-french.html",
  },
  dest: {
    ext: ".html",
    path: "/fr/other-page-french",
  },
  src: {
    asset: undefined,
    ext: ".njk",
    path: "/other",
    remote: undefined,
    slug: "other",
  },
}
`;

snapshot[`multilanguage plugin 8`] = `
{
  content: "<!DOCTYPE html>
" +
    '<html lang="it"><head><link rel="alternate" hreflang="en" href="http://localhost/en/page-1/">
' +
    '<link rel="alternate" hreflang="it" href="http://localhost/it/pagina-1/">
' +
    '<link rel="alternate" hreflang="gl" href="http://localhost/gl/paxina-1/">
' +
    "</head><body><p>Pagina 1</p>
" +
    "</body></html>",
  data: {
    alternates: [
      "en",
      "gl",
      "it",
    ],
    children: true,
    content: "Pagina 1
",
    date: 1970-01-01T00:00:00.000Z,
    id: "page1",
    lang: "it",
    mergedKeys: {
      en: "object",
      fr: "object",
      gl: "object",
      it: "object",
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [],
    title: "page-it",
    url: "/it/pagina-1/",
  },
  dest: {
    ext: ".html",
    path: "/it/pagina-1/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/pages/page1",
    remote: undefined,
    slug: "page1",
  },
}
`;

snapshot[`multilanguage plugin 9`] = `
{
  content: "<!DOCTYPE html>
" +
    '<html lang="en"><head><link rel="alternate" hreflang="en" href="http://localhost/en/page-1/">
' +
    '<link rel="alternate" hreflang="it" href="http://localhost/it/pagina-1/">
' +
    '<link rel="alternate" hreflang="gl" href="http://localhost/gl/paxina-1/">
' +
    "</head><body><p>Page 1</p>
" +
    "</body></html>",
  data: {
    alternates: [
      "en",
      "gl",
      "it",
    ],
    children: true,
    content: "Page 1
",
    date: 1970-01-01T00:00:00.000Z,
    id: "page1",
    lang: "en",
    mergedKeys: {
      en: "object",
      fr: "object",
      gl: "object",
      it: "object",
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [],
    title: "page-en",
    url: "/en/page-1/",
  },
  dest: {
    ext: ".html",
    path: "/en/page-1/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/pages/page1_en",
    remote: undefined,
    slug: "page1_en",
  },
}
`;

snapshot[`multilanguage plugin 10`] = `
{
  content: "<!DOCTYPE html>
" +
    '<html lang="gl"><head><link rel="alternate" hreflang="en" href="http://localhost/en/page-1/">
' +
    '<link rel="alternate" hreflang="it" href="http://localhost/it/pagina-1/">
' +
    '<link rel="alternate" hreflang="gl" href="http://localhost/gl/paxina-1/">
' +
    "</head><body><p>Páxina 1</p>
" +
    "</body></html>",
  data: {
    alternates: [
      "en",
      "gl",
      "it",
    ],
    children: true,
    content: "Páxina 1
",
    date: 1970-01-01T00:00:00.000Z,
    id: "page1",
    lang: "gl",
    mergedKeys: {
      en: "object",
      fr: "object",
      gl: "object",
      it: "object",
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    tags: [],
    title: "page-gl",
    url: "/gl/paxina-1/",
  },
  dest: {
    ext: ".html",
    path: "/gl/paxina-1/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/pages/page1_gl",
    remote: undefined,
    slug: "page1_gl",
  },
}
`;

snapshot[`multilanguage plugin 11`] = `
{
  content: "<!DOCTYPE html>
" +
    '<html lang="gl"><head><link rel="alternate" hreflang="en" href="http://localhost/en/page-1/">
' +
    '<link rel="alternate" hreflang="gl" href="http://localhost/gl/page-1/">
' +
    "</head><body><nav>
" +
    "
" +
    '  <a href="/en/page-1/" hreflang="en" class="alternate"></a>
' +
    "
" +
    '  <a href="/gl/page-1/" hreflang="gl" class="alternate"></a>
' +
    "
" +
    "</nav>
" +
    "
" +
    "  <li>Ola mundo</li>
" +
    "
" +
    "  <li>page-gl</li>
" +
    "
" +
    "</body></html>",
  data: {
    alternates: [
      "en",
      "gl",
    ],
    children: false,
    content: undefined,
    date: 1970-01-01T00:00:00.000Z,
    id: "page-1",
    lang: "gl",
    layout: "layout.njk",
    mergedKeys: {
      en: "object",
      fr: "object",
      gl: "object",
      it: "object",
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    pagination: {
      next: null,
      page: 1,
      previous: null,
      totalPages: 1,
      totalResults: 2,
    },
    results: 2,
    search: Search {},
    tags: [],
    url: "/gl/page-1/",
  },
  dest: {
    ext: ".html",
    path: "/gl/page-1/index",
  },
  src: {
    asset: undefined,
    ext: ".tmpl.js",
    path: "/pagination[0]",
    remote: undefined,
    slug: "pagination",
  },
}
`;

snapshot[`multilanguage plugin 12`] = `
{
  content: "<!DOCTYPE html>
" +
    '<html lang="en"><head><link rel="alternate" hreflang="en" href="http://localhost/en/page-1/">
' +
    '<link rel="alternate" hreflang="gl" href="http://localhost/gl/page-1/">
' +
    "</head><body><nav>
" +
    "
" +
    '  <a href="/en/page-1/" hreflang="en" class="alternate"></a>
' +
    "
" +
    '  <a href="/gl/page-1/" hreflang="gl" class="alternate"></a>
' +
    "
" +
    "</nav>
" +
    "
" +
    "  <li>Hello world</li>
" +
    "
" +
    "  <li>Other</li>
" +
    "
" +
    "  <li>page-en</li>
" +
    "
" +
    "</body></html>",
  data: {
    alternates: [
      "en",
      "gl",
    ],
    children: false,
    content: undefined,
    date: 1970-01-01T00:00:00.000Z,
    id: "page-1",
    lang: "en",
    layout: "layout.njk",
    mergedKeys: {
      en: "object",
      fr: "object",
      gl: "object",
      it: "object",
      tags: "stringArray",
    },
    page: undefined,
    paginate: [Function: paginate],
    pagination: {
      next: null,
      page: 1,
      previous: null,
      totalPages: 1,
      totalResults: 3,
    },
    results: 3,
    search: Search {},
    tags: [],
    url: "/en/page-1/",
  },
  dest: {
    ext: ".html",
    path: "/en/page-1/index",
  },
  src: {
    asset: undefined,
    ext: ".tmpl.js",
    path: "/pagination[1]",
    remote: undefined,
    slug: "pagination",
  },
}
`;
